src/Microsoft.VisualStudio.Services.Agent/AgentServer.cs

```diff
@@ -41,44 +54,120 @@ public interface IAgentServer : IAgentService
 
     public sealed class AgentServer : AgentService, IAgentServer
     {
-        private bool _hasConnection;
-        private VssConnection _connection;
-        private TaskAgentHttpClient _taskAgentClient;
-
-        public async Task ConnectAsync(VssConnection agentConnection)
+        private bool _hasGenericConnection;
+        private bool _hasMessageConnection;
+        private bool _hasRequestConnection;
+        private VssConnection _genericConnection;
+        private VssConnection _messageConnection;
+        private VssConnection _requestConnection;
+        private TaskAgentHttpClient _genericTaskAgentClient;
+        private TaskAgentHttpClient _messageTaskAgentClient;
+        private TaskAgentHttpClient _requestTaskAgentClient;
+
+        public async Task ConnectAsync(Uri serverUrl, VssCredentials credentials)
```

Comment:
doesnt this create unnecessary connections? should the new enum `AgentConnectionType` be `[Flags]` and exposed as a parameter to this method? That way only the required connections are created.